packages/ember-glimmer/tests/integration/syntax/each-in-test.js

```diff
@@ -0,0 +1,142 @@
+import { set } from 'ember-metal/property_set';
+import { strip } from '../../utils/abstract-test-case';
+import { applyMixins } from '../../utils/abstract-test-case';
+import { moduleFor } from '../../utils/test-case';
+import {
+  BasicConditionalsTest,
+  SyntaxCondtionalTestHelpers,
+  TruthyGenerator,
+  FalsyGenerator,
+  // ObjectTestCases
+} from '../../utils/shared-conditional-tests';
+
+class EachInTest extends BasicConditionalsTest {
+
+  get truthyValue() { return { 'Not Empty': 1 }; }
+  get falsyValue() { return {}; }
+
+}
+
+applyMixins(EachInTest,
+
+  SyntaxCondtionalTestHelpers,
+
+  new TruthyGenerator([
+    // TODO: figure out what the rest of the cases are
+    { foo: 1 }
+  ]),
+
+  new FalsyGenerator([
+    // TODO: figure out what the rest of the cases are
+    {},
+    Object.create({ 'Not Empty': 1 }),
+    undefined,
+    null
+  ])
+
+  // TODO(mmun): Add support for object proxies and
+  // include the ObjectTestCases mixin.
+);
+
+moduleFor('@htmlbars Syntax test: {{#each-in}}', class extends EachInTest {
+
+  templateFor({ cond, truthy, falsy }) {
+    return `{{#each-in ${cond}}}${truthy}{{else}}${falsy}{{/each-in}}`;
+  }
+
+  [`@test it repeats the given block for each item in the hash`]() {
+    this.render(strip`
+      <ul>
+        {{#each-in categories as |category count|}}
+          <li>{{category}}: {{count}}</li>
+        {{/each-in}}
+      </ul>
+    `, {
+      categories: {
+        'Smartphones': 8203,
+        'JavaScript Frameworks': Infinity
+      }
+    });
+
+    this.assertStableRerender();
+
+    this.assertHTML(strip`
+      <ul>
+        <li>Smartphones: 8203</li>
+        <li>JavaScript Frameworks: Infinity</li>
+      </ul>
+    `);
+
+    this.runTask(() => set(this.context, 'categories', {
+      'Smartphones': 100
+    }));
+
+    this.assertHTML(`<ul><li>Smartphones: 100</li></ul>`);
+
+    this.runTask(() => set(this.context, 'categories', {
+      'Programming Languages': 199303,
+      'Good Programming Languages': 123,
+      'Bad Programming Languages': 456
+    }));
+
+    this.assertHTML(strip`
+      <ul>
+        <li>Programming Languages: 199303</li>
+        <li>Good Programming Languages: 123</li>
+        <li>Bad Programming Languages: 456</li>
+      </ul>
+    `);
```

Comment:
This is testing replacement, which is usually covered implicitly with the "reset via replacement" step. It seems fine to also test this as an extra step in here if you want. However, can you still add the "reset via replacement" step?

```javascript
this.runTask(() => {
  set(this.context, 'categories', {
    'Smartphones': 8203,
    'JavaScript Frameworks': Infinity
  }
});

this.assertHTML(strip`
  <ul>
    <li>Smartphones: 100</li>
    <li>JavaScript Frameworks: Infinity</li>
  </ul>
`);
```

This step has helped us catch some subtle caching bugs in the past. Since it's hard to reason about when those kind of bugs are possible (we were definitely surprised a few times), we decided to make this mandatory. (And we need to test replacement anyway, which is why we usually fold it into a single "reset with replacement" step.)